### PR TITLE
Add improved search command

### LINKED ISSUE #12: Add improved find/search feature
As a user, I can search for a client by their name and attributes

### PR DESCRIPTION
**Pull request by wyrchris**
### Description

Added Search Command along with respective Parser class and Predicate class to update Model with search result. ArgumentMultiMap class has also been updated to allow for proper equality check.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update
- [ ] Code quality improvements

### How to test
Search Command format: `search KEYWORDS [MORE_KEYWORDS]... [ATTRIBUTE/ATTRIBUTE_KEYWORD]...`
Using the default/sample addressbook (delete tp/data/addressbook.json file before running the addressbook if not using the default addressbook)
Sample inputs

1. search alex yu li e/example.com t/friends will return Alex and Bernice
2. search * will just return back the entire list
3. search * a/Serangoon will return Bernice and David
4. search a p/9 e/example.com t/family will return David
5. search * t/friends a/Ang Mo Kio p/91 e/example.com will return no one

### Checklist
- [X] I have tested this code
- [X] I have updated the documentation

Closes #12

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -152,32 +152,50 @@ Format: `list`
...
-* `KEYWORD` and `MORE_KEYWORDS` will be used to match will all key
-* `attribute/` refers to the argument tag for the client's attribute
-* `ATTRIBUTE_KEYWORD` refers to the keyword that is to be matched with the corresponding client attribute
-* The search is case-insensitive. e.g `keith` will match `Keith`
-* The order of the keywords does not matter. e.g. `John Doe` will match `Doe John`
+* `KEYWORD` and `MORE_KEYWORDS` will be used to match will all attribute of the person
+* `attribute/` refers to the argument tag for the client's attribute.
+* `ATTRIBUTE_KEYWORD` refers to the keyword that is to be matched with the corresponding client attribute.
+* `*` can be used for the `KEYWORD` along with 1 or more `ATTRIBUTE/ATTRIBUTE_KEYWORD` to search using only attribute.
+* The search is case-insensitive. e.g `keith` will match `Keith`.
```
Comment (CrownKira) [inline docs/UserGuide.md:189]: It is not case-insensitive: `search Serangoon` works but not `search serangoon`

File: docs/UserGuide.md [hunk 2]
```diff
@@ -96,15 +96,15 @@ Disposable Income | Integer | disposable-income/
...
 
 * A client must have minimally the name and email tag filled during creation
 * Any other tags are optional
 * Tags that can be added are as seen in the client information in the Client Info Section
 
 Examples:
-* `create Benedict Chua /email benchua@hotmail.com`
-* `create Keith /email keithtan@ymail.com /phone-no 12345678 /risk-appetite 4`
+* `create Benedict Chua email/benchua@hotmail.com`
+* `create Keith email/keithtan@ymail.com phone-no/12345678 risk-appetite/4`
```
Comment (CrownKira) [inline docs/UserGuide.md:107]: The current implementation doesn't support full prefix email/, perhaps you should change this back to e/ (update this again in future PR when this is supported)

Comment (CrownKira) [reply docs/UserGuide.md:107]: OR perhaps you want to create an issue on this

**Review by CrownKira — CHANGES_REQUESTED** [review]
Nice~

File: docs/UserGuide.md (hunk 1 above)
Comment (wyrchris) [reply docs/UserGuide.md:189]: Have committed change to make the search case insensitive

File: docs/UserGuide.md (hunk 2 above)
Comment (wyrchris) [reply docs/UserGuide.md:107]: The current version of the user guide is already using this new syntax and the implementation of these prefix if I am not wrong will be done under the updating of client info

**Review by CrownKira — APPROVED** [review]
Seems to work. LGTM!

**Pull request merged**